Blog post for moving data between JBOD disks using Cruise Control (#469)
strimzi/strimzi.github.io: _posts/2025-01-15-moving-data-between-JBOD-disks-using-cruise-control.md

---
layout: post
title: "Moving data between JBOD disks using Cruise Control"
date: 2025-01-15
author: shubham_rawat
---

> Apache Kafka is a platform which provides durability and fault tolerance by storing messages on disks and JBOD storage is one of the storage configuration types supported by Strimzi.
> The JBOD data storage configuration allows Kafka brokers to make use of multiple disks.

Reviewer: There seems to be a bit of a disconnect / big jump in this sentence. Maybe you can lead it out with somehting like this?

```suggestion
Apache Kafka is a platform that provides durability and fault tolerance by storing messages on persistent volumes.
In most cases, each Kafka broker will use one persistent volume.
However, it is also possible to use multiple volumes for each broker.
This configuration is called JBOD storage.
```

Using JBOD storage, you can increase the data storage capacity for Kafka nodes, which can further lead to performance improvements.

> If you plan to remove a disk that contains partition replicas, the data must be safely moved to other disks first.

Reviewer: You talk about _Moving data between JBOD disks_. So it would be good to include also adding volumes  -even if briefly in one or two sentences and link to the docs.

```suggestion
It might happen that you need to add or remove volumes to increase or shrink the overall capacity and performance of the Kafka cluster.
When adding new volumes, you first need to add the volume and then move some of the data to to.
That can be done using the _intrabroker_ rebalance: TOTO Link to docs.
When removing volumes, you have to first safely move the data to other volumes first.
```

Failing to do so could result in data loss.
Moving data between the JBOD disks can be done using the `kafka-reassign-partitions.sh` tool which is not very user-friendly, therefore in Strimzi 0.45.0 we introduced the ability to move data between the JBOD disks using Cruise Control.

> ### Cruise Control to move data between JBOD disks

Reviewer: Following on the comment above - I don't think we want to go into details for adding volumes and focus on removing them for the rest of the blog post. So I would adjust the title accordingly.

        },
        {
          "partitions": [
            {
              "partition": "my-topic-2",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            }
          ],
          "error": null,
          "logDir": "/var/lib/kafka/data-2/kafka-log3"
        },
        {
          "partitions": [
            {
              "partition": "my-topic-1",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            }
          ],
          "error": null,
          "logDir": "/var/lib/kafka/data-1/kafka-log3"
        }
      ]
    }
  ],
  "version": 1
}
```

Next, let's move the data from volumes 1 and 2 to volume 0 for all the brokers. For doing that let's create a `KafkaRebalance` resource with `remove-disks` mode.
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
# no goals specified, using the default goals from the Cruise Control configuration
spec:
  mode: remove-disks
  moveReplicasOffVolumes:
    - brokerId: 3
      volumeIds: [1, 2]
    - brokerId: 4
      volumeIds: [1, 2]
    - brokerId: 5

>       volumeIds: [1, 2]

Reviewer: MAybe you can use the same YAML formatting as you have below in the output from Kubernetes? It might help to avoid confusion between the meaning of
```
volumeIds: [1, 2]
```
and 
```
      volumeIds:
        - 1
        - 2
```
among less experienced users.

data-0-my-cluster-broker-5       Bound    pvc-35c1b8ad-d4cc-4785-848a-670dc62035c3   100Gi      RWO            standard       63m
data-0-my-cluster-controller-0   Bound    pvc-7719eb5d-59d1-4e47-9fed-7ffeff566d1f   100Gi      RWO            standard       63m
data-0-my-cluster-controller-1   Bound    pvc-b1bccbe2-cd0b-4ae6-a6d7-54f2dd9728bf   100Gi      RWO            standard       63m
data-0-my-cluster-controller-2   Bound    pvc-a7b8306a-f9c3-4e62-9201-f9cb189fc3f8   100Gi      RWO            standard       63m
data-1-my-cluster-broker-3       Bound    pvc-0aa2f480-6cb8-4247-b0f7-322bb95b0b2a   100Gi      RWO            standard       63m
data-1-my-cluster-broker-4       Bound    pvc-870755ad-3ee2-4c7a-8c03-c20502014548   100Gi      RWO            standard       63m
data-1-my-cluster-broker-5       Bound    pvc-0a291b63-955b-40db-98e7-eef25224c78a   100Gi      RWO            standard       63m
data-2-my-cluster-broker-4       Bound    pvc-3d79414f-e227-4555-8589-41893d433d8d   100Gi      RWO            standard       63m
data-2-my-cluster-broker-5       Bound    pvc-0c126dc5-863f-4e2a-96ae-1a4fef9d8839   100Gi      RWO            standard       63m
```

> It is because they are not deleted by default, and you need to remove them yourself. You can delete the PVC's using the following command.

Author: If we are showing an example where we are trying to remove a JBOD disk, then yes but if we are just showing how to use the endpoint, then no

---

apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/node-pools: enabled
    strimzi.io/kraft: enabled
spec:
  kafka:
    version: 3.9.0
    metadataVersion: 3.9-IV0
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
  entityOperator:
    topicOperator: {}
    userOperator: {}
  cruiseControl: {}
```

Once the Kafka cluster is ready, now you can create some topics so that the volumes on brokers will contain some partition replicas.
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 3
  replicas: 3
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```

Let's see the partition replicas assigned to the volumes on the brokers using the `kafka-log-dir.sh` tool.
```shell

> kubectl exec -n myproject -ti my-cluster-pool-a-0   /bin/bash -- bin/kafka-log-dirs.sh --describe --bootstrap-server my-cluster-kafka-bootstrap:9092  --broker-list 3,4,5 --topic-list my-topic

Reviewer: It should be some helper Pod and not one of the Kafka nodes 😉

> kubectl exec -n myproject -ti my-cluster-pool-a-0   /bin/bash -- bin/kafka-log-dirs.sh --describe --bootstrap-server my-cluster-kafka-bootstrap:9092  --broker-list 3,4,5 --topic-list my-topic

Author: Okay, then I will add an extra step on deploying the helper pod and then this step

> kubectl exec -n myproject -ti my-cluster-pool-a-0   /bin/bash -- bin/kafka-log-dirs.sh --describe --bootstrap-server my-cluster-kafka-bootstrap:9092  --broker-list 3,4,5 --topic-list my-topic

Reviewer: I think you should be able to do just something like this:
```
kubectl -n kafka run kafka-consumer -ti --image=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0 --rm=true --restart=Never -- bin/kafka-log-dirs.sh --describe --bootstrap-server my-cluster-kafka-bootstrap:9092  --broker-list 3,4,5 --topic-list my-topic
```

---
layout: post
title: "Moving data between JBOD disks using Cruise Control"
date: 2025-01-15
author: shubham_rawat
---

Apache Kafka is a platform that provides durability and fault tolerance by storing messages on persistent volumes.
In most cases, each Kafka broker will use one persistent volume.
However, it is also possible to use multiple volumes for each broker.
This configuration is called JBOD storage.
Using JBOD storage, you can increase the data storage capacity for Kafka nodes, which can further lead to performance improvements.
It might happen that you need to add or remove volumes to increase or shrink the overall capacity and performance of the Kafka cluster.
When adding new volumes, you first need to add the volume and then move some data to it.
That can be done using the [_intrabroker_](https://strimzi.io/docs/operators/in-development/deploying#con-rebalance-str) rebalance.
When removing volumes, you have to first safely move the data to other volumes first.
Failing to do so could result in data loss.

> Moving data between the JBOD disks can be done using the `kafka-reassign-partitions.sh` tool, which is not very user-friendly, therefore in Strimzi 0.45.0 we introduced the ability to move data between the JBOD disks using Cruise Control.

Reviewer: Maybe check with @PaulRMellor ... but I wonder if someting like this would sound better.

```suggestion
Moving data between the JBOD disks can be done using the `kafka-reassign-partitions.sh` tool, which is not very user-friendly.
Therefore - in Strimzi 0.45.0 - we introduced the ability to move data between the JBOD disks using Cruise Control.
```

Reviewer 2: Looks good

> ### New `remove-disks` mode in KafkaRebalance

Reviewer: If you check the preview, the formatting looks bad due to some bad styles I guess. Maybe you can just leave out the formatting?

```suggestion
### New remove-disks mode in KafkaRebalance
```

kind: KafkaRebalance
metadata:
  creationTimestamp: "2025-01-15T10:20:37Z"
  generation: 1
  labels:
    strimzi.io/cluster: my-cluster
  name: my-rebalance
  namespace: myproject
  resourceVersion: "1755"
  uid: 335ee6b2-59d2-4326-a690-00278d125abd
spec:
  mode: remove-disks
  moveReplicasOffVolumes:
    - brokerId: 3
      volumeIds:
        - 1
        - 2
    - brokerId: 4
      volumeIds:
        - 1
        - 2
    - brokerId: 5
      volumeIds:
        - 1
        - 2
status:
  conditions:
    - lastTransitionTime: "2025-01-15T10:20:52.882813259Z"
      status: "True"
      type: ProposalReady
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 135
    numLeaderMovements: 0
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 100
    onDemandBalancednessScoreBefore: 0
    provisionRecommendation: ""
    provisionStatus: UNDECIDED
    recentWindows: 1
  sessionId: 028a7dc8-8f6d-485e-8580-93225528b587
```

> Now you can use the `approve` annotation to apply the generated proposal.

Reviewer: Should we link to the docs for more details about the approval? People wh try it for the first time might find that useful.
